source/ide/ide_histories.bas

```diff
@@ -0,0 +1,172 @@
+'Add an entry in the top position of the specified history. If the entry
+'already exists in the history, then it's just moved back to top again.
+'---------------------------------------------------------------------
+SUB AddToHistory (which$, entry$)
+    SELECT CASE which$
+        CASE "RECENT"
+            e$ = RemoveDoubleSlashes$(entry$)
+            bh% = OpenBuffer%("I", RecentFile$) 'load and/or set pos (back) to start
+        CASE "SEARCH"
+            e$ = entry$
+            bh% = OpenBuffer%("I", SearchedFile$) 'load and/or set pos (back) to start
+    END SELECT
+    lc% = 0: ue$ = UCASE$(e$)
+    WHILE NOT EndOfBuf%(bh%)
+        be$ = ReadBufLine$(bh%): lc% = lc% + 1
+        IF UCASE$(be$) = ue$ OR lc% = 100 THEN 'already known or limit reached?
+            nul& = SeekBuf&(bh%, -LEN(BufEolSeq$(bh%)), SBM_BufCurrent) 'back to prev (just read) line
+            nul& = SeekBuf&(bh%, 0, SBM_LineStart) 'and then ahead to the start of it
+            DeleteBufLine bh%
+            EXIT WHILE
+        END IF
+    WEND
+    nul& = SeekBuf&(bh%, 0, SBM_BufStart) 'rewind
+    WriteBufLine bh%, e$ 'put new (or known) in 1st position (again)
+    UpdateHistory which$
```

Comment:
I would think we want all of `AddToHistory` to happen while the lock is acquired. We could read/sync, modify. and write back all while holding the lock, ensuring that the in-memory version isn't out-of-date with the file when we make the modification.

Reply:
AddToHistory itself is a buffer operation, UpdateHistory is called to do the locked buffer2file update.

Reply:
I get the thinking, the issue I see is that if two instances end up in `AddToHIstory` at the same time then one of the history changes will be lost. If the lock is acquire before updating the in-memory copy then that issue can't happen.

Reply:
As each and every history change is directly connected to any kind of user interaction such as open a file, enter a search word or select "Clean history" from the menu, there can't be 2 or more instances fighting against each other. Even if you prepare everything and perfectly arrange the IDE windows so you just need to click to the next window and press enter to trigger such an action, I doubt you're faster then update process.